### PR TITLE
refactor(doc): improvements to hacking.rst

### PR DESCRIPTION
**Pull request by Alizter**
I have:

- Made sure line wrapping is consistent.
- Added a table of contents.
- Added information about Nix environments for devs.

### CODE REVIEW

File: doc/hacking.rst [hunk 1]
```diff
@@ -88,42 +92,68 @@ For integration tests, we use a system similar to `Cram tests
...
+running ``nix develop`` in the root of the Dune repository.
+
+Note that Dune only takes OCaml as a dependency and the rest of the dependencies
+are used when running the test suite.
+
+Running ``nix develop`` can take a while the first time, therefore it is
+advisable to save the state in a profile.
+
+```sh
+nix develop --profile /nix/profiles/dune
```
[the remark this answers is not in the document]
Comment (Alizter) [reply doc/hacking.rst:135]: Should we add `nix` to `.gitignore` then?

[the remark this answers is not in the document]
Comment (rgrinberg) [reply doc/hacking.rst:135]: no, we'll want to keep our nix library code there.

[the remark this answers is not in the document]
Comment (Alizter) [reply doc/hacking.rst:135]: What about `nix/profiles`?

[the remark this answers is not in the document]
Comment (rgrinberg) [reply doc/hacking.rst:135]: That one should be ignored because profiles shouldn't be committed

[the remark this answers is not in the document]
Comment (Alizter) [reply doc/hacking.rst]: Ah it already is ignored, my bad.

File: doc/hacking.rst [hunk 2]
```diff
@@ -18,8 +22,8 @@ file is automatically updated during development when we modify the ``dune``
 files in the repository. ``boot/duneboot.ml`` itself is built with a single
 invocation of ``ocamlopt`` or ``ocamlc`` via the ``bootstrap.ml`` ocaml script.
```
Comment (christinerose) [inline doc/hacking.rst:23]:
```suggestion
invocation of ``ocamlopt`` or ``ocamlc`` via the ``bootstrap.ml`` OCaml script.
```

File: doc/hacking.rst [hunk 3]
```diff
@@ -88,42 +92,68 @@ For integration tests, we use a system similar to `Cram tests
...
+Other times, we want to add a reproduction case for a bug reported by a user to
+ensure future changes won't reintroduce the faulty behaviour. Just like when
+programming, we turn such an idea into code, which is a formal language that a
+computer can understand. While another person reading this code might be able to
+follow and understand what the code does step by step, it isn't clear that
+they'll be able to reconstruct the original developer's idea. Even worse, they
+might understand the code in a completely different way, which would lead them
+to update it incorrectly.
+
+Setting up your development environment using Nix
```
Comment (christinerose) [inline doc/hacking.rst:122]:
```suggestion
Setting Up Your Development Environment Using Nix
```

**Review by christinerose — APPROVED** [review]
Some minor suggestions in formatting/grammar, but I'm approving it so you don't have to wait on me to approve before merging. 😄

**Review by rgrinberg — APPROVED** [review]

**Pull request merged**